Add reverse traversal

lgfa29 opened the pull request:
This PR implements reverse tree traversal. 

It introduces a new type of `Iterator` called `ReverseIterator` that can move through a tree in reverse order using the `Previous()` and `SeekReverseLowerBound()` methods. 

It also adds a new `WalkBackwards()` method to `Node` that works the same way as `Walk()` but in reverse, going from the largest to the lowest values.

banks commented on reverse_iter.go:
```diff
@@ -0,0 +1,177 @@
+		// we recurse)
+		m := len(n.edges)
+		ri.i.stack = append(ri.i.stack, n.edges[:m-1])
+		return ri.i.recurseMin(n.edges[m-1].node)
```
How does this work since we are recursing to the Min of the child not the Max? Surely that doesn't end up at the global max once we have recursed down multiple layers?

lgfa29 replied:
Good catch! It's a copy-paste mistake...it should `recurseMax`. I will implement the fuzzying test and see if it catches this.

banks commented on reverse_iter_test.go:
```diff
@@ -0,0 +1,148 @@
+
+func TestReverseIterator_SeekReverseLowerBound(t *testing.T) {
+	r := New()
+	keys := []string{"001", "002", "005", "010", "200", "201"}
```
This is a good basic test but I suggest implementing a fuzz test to exercise all the edge cases more thoroughly.

I _guess_ that this case is not actually creating a tree with more than a single node depth which possibly explains why the apparent bug i noted isn't being triggered.

Of course I could just be wrong about that bug, but regardless I found _lots_ of bugs in the original LowerBound implementation from `TestIterateLowerBoundFuzz` that a sinple test like this would not have picked up.

lgfa29 replied:
That's a nice test! I somehow missed it among the others, but I will add that and see if catches the bug you found.

banks reviewed (changes requested):
Looking really good!

I suggest a fuzz test to exercise more of the edges and possibly found a bug but I could be wrong!

lgfa29 commented:
@banks I added fuzz tests for the `ReverseIterator`. It was super helpful, so thanks for the tip.

I noticed that the recurse bug is only triggered when the seek starts from a non-root node, so I added two tests. There's a comment explaining why this is the case. I hope the explanation is clear enough, otherwise I can expand it. It's kind of an unusual case, but I guess not impossible, so worth testing.

The two tests are very similar, but they are different enough that I couldn't find a nice way to re-use things, I hope that's OK.

banks commented on reverse_iter_test.go:
```diff
@@ -0,0 +1,224 @@
+	// tree (with a null byte appended since our tree doesn't support one key
+	// being a prefix of another and treats null bytes specially).
+	//
+	// It also maintains a plain sorted list of the same set of keys and asserts
+	// that iterating from some random key to the end using LowerBound produces
+	// the same list as filtering all sorted keys that are lower.
```
Is this comment right? Seems like copy paste from the normal direction perhaps we could change it to make it make sense for the reverse direction?

lgfa29 replied:
Yes...again...thank you 😄

banks reviewed (approved):
Great job Luiz!

I think it could be possible to work the two fuzz tests into one with some random bool input that chooses wether or not to try to start at a non-root node. Then both paths would be exercised with high probability in any given run (check runs 1000 iterations by default IIRC).

That said, I don't think it's especially problematic to have them separate either so accepting for now.
